File: modules/uml/app/model/UmlExercise.java
package model;

import java.util.List;
import java.util.Map;
import java.util.Set;

import javax.persistence.Column;
import javax.persistence.Entity;

import com.fasterxml.jackson.annotation.JsonGetter;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSetter;

import io.ebean.Finder;
import io.ebean.annotation.DbJson;
import model.exercise.Exercise;

@Entity
public class UmlExercise extends Exercise {
  
  public static final Finder<Integer, UmlExercise> finder = new Finder<>(UmlExercise.class);
  
  @Column(columnDefinition = "text")
  @JsonIgnore
  public String classSelText;
  
  @Column(columnDefinition = "text")
  @JsonIgnore
  public String diagDrawText;
  
  @Column(columnDefinition = "text")
  @JsonProperty(required = true)
  public String solution;
  
  @DbJson
  @JsonIgnore
  public Map<String, String> mappings;
  
  @DbJson
  public List<String> ignoreWords;
  
  public UmlExercise(int id) {
    super(id);
  }
  
  @JsonIgnore
  public String getClassesForDiagDrawingHelp() {
    return UmlSolution$.MODULE$.getClassesForDiagDrawingHelp(getSolution().classes());
  }
  
  @JsonGetter("mappings")
  @JsonProperty(required = true)
  public Set<Map.Entry<String, String>> getMappingsForJson() {
    return mappings.entrySet();
  }
  
  @JsonIgnore
  public UmlSolution getSolution() {
    return UmlSolution.fromJson(solution).get();
  }
  
  @JsonGetter("solution")
  public Object getSolutionAsJson() {
    // Setter only for generation of json schema...
    return solution;
  }
  
  @JsonSetter("solution")
  public void setSolution(Object theSolution) {
    // Getter only for generation of json schema...
    solution = theSolution.toString();
  }
  
}

File: modules/uml/app/model/UmlExercise.scala
package model

import java.nio.file.Path
import javax.persistence.{Column, Entity}

import com.fasterxml.jackson.annotation.{JsonGetter, JsonIgnore, JsonProperty, JsonSetter}
import io.ebean.Finder
import io.ebean.annotation.DbJson
import model.exercise.Exercise
import play.twirl.api.{Html, HtmlFormat}

import scala.util.Try

@Entity
class UmlExercise extends Exercise {

  @Column(columnDefinition = "text")
  @JsonIgnore
  var classSelText: String = _

  @Column(columnDefinition = "text")
  @JsonIgnore
  var diagDrawText: String = _

  @Column(columnDefinition = "text")
  @JsonProperty(required = true)
  var solution: String = _

  @DbJson
  var mappings: java.util.Map[String, String] = _

  @DbJson
  @JsonProperty(value = "ignoreWords", required = true)
  var ignoreWords: java.util.List[String] = _

  override def renderRest(fileResults: List[Try[Path]]): Html = new Html(
    s"""<td>${views.html.helperTemplates.modal.render("Klassenwahltext...", new Html(classSelText + "<hr>" + HtmlFormat.escape(classSelText)), "Klassenwahltext")}</td>
       |<td>${views.html.helperTemplates.modal.render("Diagrammzeichnentext...", new Html(diagDrawText + "<hr>" + HtmlFormat.escape(diagDrawText)), "Diagrammzeichnentext")}</td>""".stripMargin)

  @JsonIgnore
  def getClassesForDiagDrawingHelp: String = UmlSolution.getClassesForDiagDrawingHelp(getSolution.classes)

  //  @JsonGetter("mappings")
  @JsonProperty(value = "mappings", required = true)
  def getMappingsForJson: java.util.Set[java.util.Map.Entry[String, String]] = mappings.entrySet

  @JsonIgnore
  def getSolution: UmlSolution = UmlSolution.fromJson(solution).get

  @JsonGetter("solution")
  def getSolutionAsJson: Object = {
    // Setter only for generation of json schema...
    solution
  }

  @JsonSetter("solution")
  def setSolution(theSolution: Object) {
    // Getter only for generation of json schema...
    solution = theSolution.toString
  }

}

object UmlExercise {

  val finder: Finder[Integer, UmlExercise] = new Finder(classOf[UmlExercise])
}